### PR TITLE
expectations: pass num hermite points to _expectation() and minor cleaning

### PR DESCRIPTION
**Pull request by ialong**
This pull request fixes two issues:

- the signature `(Gaussian, identity_mean, None, kernel, features)` was calling itself whenever no specific implementation for `kernel` exists because `Identity` is a type of `Linear` mean function
- the default number of Hermite points (often too high) was used for nested calling of `expectation` as the `num_gauss_hermite_points` argument was not being passed beyond the first call

What is irking about the current fix for the second issue:

- `_expectation`'s should, in theory, have no knowledge of Hermite points and should not receive them as arguments
- the code is now slightly more verbose, though at least explicit
- to keep the code and signatures simple (and not changing 35 of them) the number of Hermite points argument (`nghp`) is not checked by multiple dispatch

The solution to fix all this would be to make `expectation` into a class but it is unclear at the moment whether this would have any other use cases. The Hermite points issue does not warrant, in my opinion, changing `expectation` to a class as that would imply a significant refactor and changes to code outside of the `expectations` submodule.

### CODE REVIEW

File: gpflow/expectations.py [hunk 1]
```diff
@@ -171,14 +171,14 @@ def _quadrature_expectation(p, obj1, feature1, obj2, feature2, num_gauss_hermite
 
 # =========================== ANALYTIC EXPECTATIONS ===========================
 
-def expectation(p, obj1, obj2=None, num_gauss_hermite_points=None):
+def expectation(p, obj1, obj2=None, nghp=None):
```
Comment (awav) [inline gpflow/expectations.py]: :param ...: description
:return: description

follow _reStructuredText_ (reST, sphinx) docstrings 
https://pythonhosted.org/an_example_pypi_project/sphinx.html

File: gpflow/expectations.py [hunk 2]
```diff
@@ -571,37 +588,33 @@ def _expectation(p, linear_mean, none, kern, feat):
...
+    so _expectation(p, linear_mean, none, kern, feat) would call itself.
+    More specific signatures (e.g. (p, identity_mean, None, RBF, feat)) will
+    be found and used whenever available
     """
-    with params_as_tensors_for(constant_mean):
-        c = constant_mean(p.mu)  # NxQ
-        eKxz = expectation(p, (kern, feat))  # NxM
-
-        return c[..., None] * eKxz[:, None, :]
+    raise NotImplementedError
```
Comment (awav) [inline gpflow/expectations.py]: So, other cases can't cause infinite loop? Did you check it? Can you prove it?

Comment (vdutor) [reply gpflow/expectations.py]: I think it is definitely worth adding @stefanosele's case to the unit tests and show that this is remedied now.

Comment (ialong) [reply gpflow/expectations.py]: Yes I have checked there are no other self-referential calls to `expectation`. Not sure a test is absolutely necessary

Comment (ialong) [reply gpflow/expectations.py]: Ok I added a simple one by including a Matern kernel in exKxz expectation tests

File: gpflow/expectations.py (hunk 1 above)
Comment (ialong) [reply gpflow/expectations.py]: I have changed the docstrings, do they look ok now?

File: gpflow/expectations.py (hunk 2 above)
Comment (stefanosele) [reply gpflow/expectations.py]: I'm happy with that

**Comment by stefanosele** [comment]
Good! Finally someone took the decision for  passing around kwargs vs. nghp.

Just close https://github.com/GPflow/GPflow/pull/693 and we're good to go

**Pull request merged**